### ISSUE TITLE
"IPython cluster could not be started. Giving up"

### ISSUE THREAD

**Issue by DrAndiLowe** [fhirschmann/knitron#3]
This is a duplicate of the following, which hasn't been resolved:
http://stackoverflow.com/questions/30306661/using-ipython-engine-in-knitr

I see similar behaviour, and knitron doesn't work for me:

Starting cluster for profile knitr
sh: 1: ipcluster: not found
Quitting from lines 35-36 (test.Rmd) 
Error in knitron.start(profile) : 
  IPython cluster could not be started. Giving up.
In addition: There were 42 warnings (use warnings() to see them)

I tried this, using your own README.Rmd file in this package:
library(knitron)
setwd("~/Documents/PyMarkdown")
knit("README.Rmd")

Clicking on the "knit HTML" button in RStudio will produce an HTML document with text, but there is no output from the Python code chunks (no text, no plots). Additionally, this is output to the console:

output file: README.knit.md

/usr/lib/rstudio/bin/pandoc/pandoc +RTS -K512m -RTS README.utf8.md --to html --from markdown+autolink_bare_uris+ascii_identifiers+tex_math_single_backslash --output README.html --smart --email-obfuscation none --self-contained --standalone --section-divs --template /home/andy/R/i686-pc-linux-gnu-library/3.2/rmarkdown/rmd/h/default.html --variable 'theme:bootstrap' --include-in-header /tmp/RtmpfTA1Pn/rmarkdown-str4fa212dea82e.html --mathjax --variable 'mathjax-url:https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML' --no-highlight --variable highlightjs=/home/andy/R/i686-pc-linux-gnu-library/3.2/rmarkdown/rmd/h/highlight --variable navigationjs=/home/andy/R/i686-pc-linux-gnu-library/3.2/rmarkdown/rmd/h/navigation-1.0 

Output created: README.html
Warning messages:
1: In get_engine(options$engine) :
  Unknown language engine 'ipython' (must be registered via knit_engines$set()).
2: In get_engine(options$engine) :
  Unknown language engine 'ipython' (must be registered via knit_engines$set()).
3: In get_engine(options$engine) :
  Unknown language engine 'ipython' (must be registered via knit_engines$set()).
4: In get_engine(options$engine) :
  Unknown language engine 'ipython' (must be registered via knit_engines$set()).
5: In get_engine(options$engine) :
  Unknown language engine 'ipython' (must be registered via knit_engines$set()).
6: In get_engine(options$engine) :
  Unknown language engine 'ipython' (must be registered via knit_engines$set()).

If I do rmarkdown::render("README.Rmd") I see this:

Traceback (most recent call last):  File "/usr/local/bin/ipython", line 9, in <module>    load_entry_point('ipython==4.0.0', 'console_scripts', 'ipython')()  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 337, in load_entry_point    return get_distribution(dist).load_entry_point(group, name)  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 2279, in load_entry_point    return ep.load()  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 1989, in load    entry = **import**(self.module_name, globals(),globals(), ['**name**'])ImportError: No module named IPython
Quitting from lines 12-13 (README.Rmd) 
Error in file(con, "r") : cannot open the connection
In addition: Warning messages:
1: running command ''ipython' /home/andy/R/i686-pc-linux-gnu-library/3.2/knitron/python/knitron.py knitr chunk /tmp/RtmphPVM5u/file4f085b08dc18 2>&1 < '/tmp/RtmphPVM5u/file4f08a86b354'' had status 1 
2: In file(con, "r") :
  cannot open file '/tmp/RtmphPVM5u/file4f085b08dc18': No such file or directory

Which seems to indicate that IPython is not installed, which is not true:

andy@GLaDOS:~$ which ipython
/home/andy/anaconda3/bin/ipython
andy@GLaDOS:~$ ipython
Python 3.5.1 |Anaconda 2.4.0 (32-bit)| (default, Dec  7 2015, 11:17:45) 
Type "copyright", "credits" or "license" for more information.

IPython 4.2.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: This looks OK...

I'm unable to install libgit2-dev. I don't know if this the the source of the problem, or unconnected.

I did also try launching ipcluster from the command line:

andy@GLaDOS:~$ ipcluster start --profile=knitr --n=1
2016-04-26 13:43:20.700 [IPClusterStart] Starting ipcluster with [daemon=False]
2016-04-26 13:43:20.703 [IPClusterStart] Creating pid file: /home/andy/.ipython/profile_knitr/pid/ipcluster.pid
2016-04-26 13:43:20.704 [IPClusterStart] Starting Controller with LocalControllerLauncher
2016-04-26 13:43:21.714 [IPClusterStart] Starting 1 Engines with LocalEngineSetLauncher
2016-04-26 13:43:51.734 [IPClusterStart] Engines appear to have started successfully

This didn't help.

I would very much appreciate this issue being resolved soon, as I am hoping to demonstrate the use of knitr + ipython to my colleagues at CERN. We've started using IPython and Jupyter notebooks for our analyses recently, and there is a push towards reproducible workflows. Being able to knit python+markdown would very interesting for my community. So hoping this can be fixed!

**Comment by fhirschmann** [comment]
Hi Andrew,

Thank you for you detailed bug report! Please first try updating knitron. If it still doesn't work, please see my remarks below.

## RStudio Usage

The errors concerning an unknown IPython engine are most likely due to knitron not registering itself as `ipython` engine. Please see [this example for RStudio usage](https://raw.githubusercontent.com/fhirschmann/knitron/master/examples/rstudio.Rmd).

## IPython cluster could not be started

This can unfortunately have many reasons. Let me explain how knitron starts up such clusters:

In [knitron.R](https://github.com/fhirschmann/knitron/blob/master/R/knitron.R#L7), the `knitron.start()` function fires up an IPython cluster using `system2` and `ipcluster start --profile=knitron --n 1` as arguments. It then tries to execute code repeatedly to see if the cluster is up and running using `knitron.is_running()`.

It would be interesting to see why this has not worked. You can start it up manually using `library(knitron);knitron.start()`. I discovered a problem that arises due to the [IPython python module freezing](https://github.com/fhirschmann/knitron/blob/master/inst/python/knitron.py#L42) when it's instantiated before the cluster is up. I added a [simple workaround](https://github.com/fhirschmann/knitron/blob/master/R/knitron.R#L17) for this problem.

However in your case, I think system2 fails to find ipcluster2. You can check whether R finds ipcluster2 by executing `system2("ipcluster")`.

You can also try starting up a cluster manually using `ipcluster start --profile=knitr --n=1`. Then, check if you can execute code by running `library(knitron);knitron.execute_code("4 + 5")` in R.

## ImportError: No module named IPython

This one is quite strange. I see you're using two python installations, Anaconda and some other installation in `/usr/local`. In either case, IPython started up in your example nicely but then failed to import the IPython python module. Your `which ipython` command points to your Anaconda distribution, while your rmarkdown call uses `/usr/local/bin/ipython`.

You can check whether your python installation used by RStudio works correctly by executing `/usr/local/bin/ipython` followed by the execution of `import IPython` (on older Python versions) or `from ipyparallel import Client` (on newer versions).

**Comment by DrAndiLowe** [comment]
Hi Fabian,

I have something that sort of works now, but only with Python 2. I haven't managed to get this to work with Python 3. I spent most of today down a rabbit hole trying to debug using the error messages I was seeing when I tried your suggestions above, but there seemed to be no easy solution to the problem I resorted to uninstalling every Python library on my system except for the absolute bare bones necessary for my system (Ubuntu 12.04 LTS). Deleted everything. Then built up again, piece-wise. Deleting my Anaconda install of Python 3 was insufficient. library(knitron); rmarkdown::render("foo.Rmd") works for me. Even so, matplotlib has some weird behaviour; plt.show() doesn't render my plots in the final document, but omitting this line does. This might have something to do with the matplotlib backend, which is TkAgg on my system. I tried variations such as Qt4Agg, GTK, GtkAgg, but this was just trial and error, and nothing worked. The plots do seem to show up, but I don't understand why the show() method causes problems. Also, my biggest problem now is that all of this works, but only with Python 2. I'd like this to work with my Anaconda install of Python 3, which is _much_ easier to manage. It took me all of day, uninstalling and reinstalling various libraries, to get something that works, and I have no idea what the "secret sauce" is that makes it work -- only that some conflict between libraries, or between Python 2 and Python 3, or their libraries, was at fault. How do I specify that knitron uses my Anaconda install of Python 3? The only reason I haven't uninstalled Python 2 is because it would break my system if I did. Else I use Python 3 exclusively, because I don't want to get bogged-down with juggling between them. Is there some way of forcing knitron to use Python 3 only? Please note, I'm only at the "hello world" level with Python, although I hope to improve! My motivation for doing this is to convince my colleagues to use knitr to manage reproducible research, as I'm presently doing (very easily, thanks to RStudio and knitr) with R. Thanks!

Best regards,

Andrew.

**Comment by DrAndiLowe** [comment]
P.S. library(knitron); knit("foo.Rmd") produces foo.md, but not html or pdf. That's why I use rmarkdown::render.

**Comment by DrAndiLowe** [comment]
Hi Fabian,

Problem solved! You can close this issue. However, you might want to update the documentation with some troubleshooting tips. Here's what I've found out, and how I got things working. I hope the information is useful.

The source of the problem seems to come from multiple installations of Python (Python 2 and my Anaconda-installed Python 3 install). I'm running Ubuntu 12.04 LTS, and there are system dependencies on Python 2, so I can't uninstall Python 2 without breaking stuff really badly. It appears that Anaconda adds a line to my .bashrc so that the Python 3 version of ipython is in my path, but this doesn't seem to get picked up when running knitron, which then seems to want to use the Python 2 version. The solution to this problem is to specify the engine.path for ipython 3 in a code chunk preamble at the beginning of the Rmd file. Mine looks something like this:

``` r setup, echo = FALSE, message = FALSE, warning = FALSE, error = FALSE
# Set globals:
require(knitron)
require(knitr)
opts_chunk$set(
  engine.path="/home/andy/anaconda3/bin/ipython",
  tidy =F,
  knitron.autoplot = "auto",
  knitron.matplotlib = T,
  knitron.print = F,
  cache = FALSE, 
  dev = 'pdf'
  )
```

I'm not entirely sure if I need all of that. The essential line is the specification of the path where ipython 3 lives. This appears to be enough to disambiguate between the two Python versions. Also, it is possible to invoke knitron from RStudio by clicking on the "knit" button -- so there's no need to run knitron from the R console. My YAML looks something like this:

---

title: "This is the title"
author:
- "Andrew John Lowe"
- "Wigner Research Centre for Physics, Hungarian Academy of Sciences"
  date: "April 26, 2016"
  abstract: |
  This is the abstract.
  
  It consists of two paragraphs.
  output:
  pdf_document:
    fig_caption: yes
    keep_tex: yes
    toc_depth: 1
  
  ## html_document: default

This may or may not be the source of the magic. I click on "knit pdf", and _presto!_ my PDF appears a few seconds later. Sometimes there are error or warning messages, but it seems these can be ignored. Sometimes it fails the first time, and I need to click again, but this is not a problem.

Sometimes I see:

Output created: demo.pdf
Warning message:
running command ''ipython' /home/andy/R/i686-pc-linux-gnu-library/3.2/knitron/python/knitron.py knitr code 0 2>&1' had status 1 
2016-04-29 16:38:42.839 [IPClusterStart] ERROR | IPython cluster: stopping
2016-04-29 16:38:42.840 [IPClusterStart] Stopping Engines...

But I get my PDF in any case, so I don't care about this error message.

One oddity that remains is that plots are only shown if plt.show() is not called. That is, the default is that the last plot and text is printed automatically. This is a bit counter-intuitive, and I'd like a bit more control over this behaviour. From looking at the code, I see the following defaults:

  knitron.autoplot = "auto",
  knitron.matplotlib = T,
  knitron.print = F,

I don't usually want to print out text unless I explicitly request to, so I set knitron.print to FALSE as a global option. The other two switches are required if I want to see any plots. Setting either of them to anything other than their default settings seems to stop my plots from appearing, even after I explicitly do plt.show() to make them appear. This is a bit odd, and I'd like to understand how to manage this behaviour better, but it's not a real problem as long as I remember that one chunk = one plot. 

I haven't tested extensively, and I'm not sure what the minimum requirements are to get the magic to happen, but so everything else seems to work as expected -- which is GREAT!

Specifying ipython 3 as my knitr engine seems to make Python (not IPython!) code chunks sometimes produce output that cannot be LaTeX'd. Some error message about utf8 encodings:

``` python
friends = ['John', 'Pat', 'Gary', 'Michael']
for i, name in enumerate(friends):
    print "iteration {iteration} is {name}".format(iteration=i, name=name)
```

Results in:

output file: demo.knit.md

/usr/lib/rstudio/bin/pandoc/pandoc +RTS -K512m -RTS demo.utf8.md --to latex --from markdown+autolink_bare_uris+ascii_identifiers+tex_math_single_backslash --output demo.pdf --template /home/andy/R/i686-pc-linux-gnu-library/3.2/rmarkdown/rmd/latex/default.tex --highlight-style tango --latex-engine pdflatex --variable graphics=yes --variable 'geometry:margin=1in' 
2016-04-29 16:44:38.679 [IPClusterStart] Engines appear to have started successfully
! Package inputenc Error: Keyboard character used is undefined
(inputenc)                in inputencoding `utf8'.

See the inputenc package documentation for explanation.
Type  H <return>  for immediate help.
 ...                                              

l.276 \end{verbatim}

Try running pandoc with --latex-engine=xelatex.
pandoc: Error producing PDF from TeX source
Error: pandoc document conversion failed with error 43
In addition: Warning message:
running command ''ipython' /home/andy/R/i686-pc-linux-gnu-library/3.2/knitron/python/knitron.py knitr code 0 2>&1' had status 1 
Execution halted
2016-04-29 16:44:42.051 [IPClusterStart] ERROR | IPython cluster: stopping
2016-04-29 16:44:42.052 [IPClusterStart] Stopping Engines...

In that case, I just specify ipython as the engine for that chunk, and the problem goes away. Although I did try different LaTeX engines (xelatex, lualatex), I didn't spend any more time than this trying to debug this, as there's an easy workaround. It's as easy as replacing `{python} with`{ipython}, and then there is no problem.

So, to summarise: all is good, except that the mechanism for producing plots in the document is a bit odd, and you might want to add some troubleshooting tips.

Thanks for your help with debugging this, and from writing such a great package!

Best regards,

Andrew.

**Comment by fhirschmann** [comment]
Hi Andrew,

Thanks for your detailed input and sorry that it was such a hassle to set up. :)

I spend some time to make it a little more robust, in particular the code that starts and stops ipcluster. You can now set your preferred ipython/ipcluster commands via:

``` r
options(ipcluster = "/home/andy/anaconda3/bin/ipcluster", ipython = "/home/andy/anaconda3/bin/ipython")
```

The logging was also improved and is now hopefully more helpful than it was before. The `engine.path` option is an option in `knitr` and not `knitron`, so I really don't know why that even worked.

The other error your observed, i.e.

```
running command ''ipython' /home/andy/R/i686-pc-linux-gnu-library/3.2/knitron/python/knitron.py knitr code 0 2>&1' had status 1 
```

is non-fatal and is due to knitron checking if IPython has finished starting up. This message should now be gone with the newest version of knitron.

Concerning the `show()` method in matplotlib: You're right, in knitron you are not supposed to call this method directly. It is used in matplotlib to draw a plot using a backend of your choice. 

In knitron, this works a little bit different. `matplotlib.pyplot` is basically a state machine (or a global variable) that holds a list of plots. Pyplot is imported in knitron as `plt`. Hence, all plotting should be done through this interface. After a chunk has completed, knitron simply [collects the most recent plot and saves it to a pdf/png file](https://github.com/fhirschmann/knitron/blob/master/inst/python/knitron.py#L146).

You are also right that it only supports one plot per chunk. This is currently a limitation in knitron.

I will continue working on the other issues. There also seems to be a problem with the correct path to figures when using RStudio.

**Comment by fhirschmann** [comment]
Concerning the unicode issues: I think the reason you experience these is that you mix ipcluster and ipython with different versions, e.g. ipcluster is running version 2 and ipython is running version 3.

In your example snippet

```
print "iteration {iteration} is {name}".format(iteration=i, name=name)
```

you use the `print` statement which does not exist in Python 3 (in Python 3 we only have the print function). So I'm assuming your ipcluster is running Python 2 whereas the python script used to communicate with the cluster (`inst/python/knitron.py`) is running version 3. I also accidentally did this just now, and I added some code that produces some warnings when this version mismatch is detected. [I added some example that uses Python 3](https://raw.githubusercontent.com/fhirschmann/knitron/master/examples/python3.Rmd). This unicode issue may still surface if you have ipcluster(2) running in the background outside the control of knitron.

**Comment by DrAndiLowe** [comment]
Ah, so I should add:

``` r, echo = F
suppressPackageStartupMessages(library(knitron))
options(ipython = "ipython3", ipcluster = "ipcluster3")
```

``` r, engine = "ipython"
import sys
sys.version
```

in a code chunk preamble at the beginning of my document? 

sh: 1: ipython3: not found

Quitting from lines 49-51 (demo.Rmd) 
Error in system2(python, c("profile", "list"), stdout = TRUE) : 
  error in running command
Calls: <Anonymous> ... in_dir -> engine -> knitron.is_running -> gsub -> system2
Execution halted

H'mm. OK, that didn't seem to work. I think there's something I'm not understanding properly here.

In any case, this isn't a big problem for me.

I still encounter the behaviour that the inclusion of plt.show() results in no plot -- I understand why (that knitron handles the mechanism for producing the plot), but perhaps it would be better if this didn't fail silently? This might trip up some people who start using knitron for the first time and find that code that is cut and pasted from their working IPython Notebook doesn't seem to work as they expect. A warning is needed, perhaps? Just a suggestion.

Thanks once again!

**Comment by fhirschmann** [comment]
Hi Andrew,

You have to include the complete path when calling `options()` when it's not in your `$PATH`. I tried it with anaconda3 and it worked for me after the following steps:
- Install ipyparallel for anaconda: `/home/andy/anaconda3/bin/pip install ipyparallel`
- Use [the example](https://raw.githubusercontent.com/fhirschmann/knitron/master/examples/python3.Rmd) and adapt the PATH as follows: `options(ipcluster = "/home/andy/anaconda3/bin/ipcluster", ipython = "/home/andy/anaconda3/bin/ipython3")`

The output should be

```
## '3.5.1 |Anaconda 4.0.0 (64-bit)| (default, Dec  7 2015, 11:16:01) \n[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]'
```

You basically have to make sure that `ipcluster` and `ipython` match, i.e. both are of the same version and ideally from the same distribution such as anaconda.

The `sys.version` that comes in a code chunk following the chunk with the option setting was just [an example](https://raw.githubusercontent.com/fhirschmann/knitron/master/examples/python3.Rmd)  that shows which version of python you are running.

Concerning the `plt.show()` issue. I will think about how to make it clear not to call this function. It probably removes the plot from the pyplot registry, leaving knitron with no plots.

**Comment by fhirschmann** [comment]
Feel free you reopen if the problem persists.

**Comment by sahilseth** [comment]
I possibly had the same (or similar issue). So am specifying the path for `ipython` and `ipcluster`.

After doing the above, I am able to run `knitron.start`

```
knitron::knitron.start()

knitron/INFO Starting cluster for profile knitr via /Users/sahilseth/apps/conda/3.5/bin/ipcluster start --profile=knitr --n=1
knitron/DEBUG Waiting for IPython engine to start up (see /var/folders/qg/wgvd9bj9151fbtk1wtn6r7km0000gp/T//RtmpF6AoY0/ipcluster_17ab5791d4706)
knitron/WARN Cluster is already running - reusing
knitron/INFO Engine started up successfully
knitron/DEBUG Executing /Users/sahilseth/apps/conda/3.5/bin/ipython /usr/local/lib/R/3.3/site-library/knitron/python/knitron.py knitr isrunning
knitron/ERROR Communication with engine for profile knitr failed
```

However I see an error in the end, suggesting that communication with the specific profile failed. And subsequent chunks do not work.

Thanks for your help.
